libs/execution_context/ebpf_core.c

```diff
@@ -894,6 +951,54 @@ _ebpf_core_protocol_map_delete_element(_In_ const ebpf_operation_map_delete_elem
     EBPF_RETURN_RESULT(retval);
 }
 
+static ebpf_result_t
+_ebpf_core_protocol_map_delete_element_batch(
+    _In_ const ebpf_operation_map_delete_element_batch_request_t* request,
+    _Inout_ ebpf_operation_map_delete_element_batch_reply_t* reply)
+{
+    EBPF_LOG_ENTRY();
+    ebpf_result_t retval;
+    ebpf_map_t* map = NULL;
+    size_t key_length;
+    size_t input_count = 0;
+    size_t output_count = 0;
+
+    retval = EBPF_OBJECT_REFERENCE_BY_HANDLE(request->handle, EBPF_OBJECT_MAP, (ebpf_core_object_t**)&map);
+    if (retval != EBPF_SUCCESS) {
+        goto Done;
+    }
+
+    retval = ebpf_safe_size_t_subtract(
+        request->header.length, EBPF_OFFSET_OF(ebpf_operation_map_delete_element_request_t, key), &key_length);
+    if (retval != EBPF_SUCCESS) {
+        goto Done;
+    }
+
+    const ebpf_map_definition_in_memory_t* map_definition = ebpf_map_get_definition(map);
+
+    if (key_length % map_definition->key_size != 0) {
+        retval = EBPF_INVALID_ARGUMENT;
+        goto Done;
+    }
+
+    input_count = key_length / map_definition->key_size;
+
+    for (output_count = 0; output_count < input_count; output_count++) {
+        retval = ebpf_map_delete_entry(
+            map, map_definition->key_size, request->keys + output_count * map_definition->key_size, 0);
+        if (retval != EBPF_SUCCESS) {
+            goto Done;
+        }
+    }
+
+    reply->header.length = (uint16_t)sizeof(ebpf_operation_map_delete_element_batch_reply_t);
+    reply->count_of_elements_processed = (uint32_t)output_count;
+
+Done:
+    EBPF_OBJECT_RELEASE_REFERENCE((ebpf_core_object_t*)map);
```

Comment:
This macro ends up calling `ebpf_object_release_reference` which calls `_update_reference_history` unconditionally _before_ checking for a null object.  This _could_ lead to inconsistent reference history if the `map` object was null (`EBPF_OBJECT_REFERENCE_BY_HANDLE` above failed for some reason).

Reply:
The issue is in ebpf_object_release_reference. If it's passed NULL it shouldn't update the history.